Teleport improvements

kontsevoy opened the pull request:
You can look into each commit separately, but here is the overview:
- Added support for arbitrary users for integration tests. You can create users separately from "teleport instances".
- Small Teleport commands like "status" or "version" do not look for web assets anymore.
- Integrated "testauthority" into integration testing
- Improved polling logic in a few places

kontsevoy committed:
Teleport won't look for web assets

... when executing simple commands like 'status' or 'version'

kontsevoy committed:
Added explicit tunnel access to integratin testing

kontsevoy committed:
Nicer inetgration package

kontsevoy committed:
FIxed the build

kontsevoy committed:
Fixed tests and made logging less verbose

kontsevoy committed:
Fixed tests

kontsevoy committed:
Dynamic ports for integration testing

kontsevoy committed:
Incorporated 'testauthority' into integration tests

kontsevoy committed:
Prepared previous commits for merging into master

- Fixed all tests
- Removed "magic constants" in random places
- Improved 'retry connecting to auth server' logic (it used to always
  fail on 1st attempt)

klizhentas commented:
there's a data race in tests:


```
time="0001-01-01T00:00:00Z" level=info msg="closing session channel" component=node fields=map[local:127.0.0.1:45992 remote:127.0.0.1:56993 login:jenkins teleportUser:jenkins id:6] file="srv/sshserver.go:600" 
time="0001-01-01T00:00:00Z" level=info msg="Asking Teleport to stop" file="_obj_test/helpers.go:419" 
time="0001-01-01T00:00:00Z" level=info msg="BroadcastEvent: event(TeleportExit)" file="service/supervisor.go:185" 
time="0001-01-01T00:00:00Z" level=info msg="client disconnected" component=proxy fields=map[local:127.0.0.1:41907 remote:127.0.0.1:37499 login:jenkins teleportUser:jenkins id:5] file="srv/sshserver.go:617" 
time="0001-01-01T00:00:00Z" level=info msg="closing session channel" component=proxy fields=map[remote:127.0.0.1:37499 login:jenkins teleportUser:jenkins id:5 local:127.0.0.1:41907] file="srv/sshserver.go:600" 
==================
WARNING: DATA RACE
Read by goroutine 86:
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint.func5()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:716 +0x88
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).onExit.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:429 +0x17a

Previous write by goroutine 83:
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:673 +0xa1c
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d

Goroutine 86 (running) created at:
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).onExit()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:431 +0x4c
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:718 +0x1527
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxy.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:577 +0x5b4
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d

Goroutine 83 (running) created at:
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:148 +0x12b
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).Register()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:109 +0x3d6
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).RegisterFunc()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:121 +0x63
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:685 +0x11e0
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxy.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:577 +0x5b4
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d
==================
==================
WARNING: DATA RACE
Read by goroutine 86:
  crypto/tls.(*listener).Close()
      <autogenerated>:35 +0x54
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint.func5()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:716 +0xa8
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).onExit.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:429 +0x17a

Previous write by goroutine 83:
  crypto/tls.Listen()
      /tmp/workdir/go/src/crypto/tls/tls.go:77 +0x177
  github.com/gravitational/teleport/lib/utils.ListenTLS()
      /gopath/src/github.com/gravitational/teleport/lib/utils/tls.go:46 +0x16f
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:673 +0x9d7
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d

Goroutine 86 (running) created at:
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).onExit()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:431 +0x4c
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:718 +0x1527
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxy.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:577 +0x5b4
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d

Goroutine 83 (running) created at:
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:148 +0x12b
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).Register()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:109 +0x3d6
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).RegisterFunc()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:121 +0x63
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:685 +0x11e0
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxy.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:577 +0x5b4
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d
==================
==================
WARNING: DATA RACE
Read by goroutine 86:
  net.(*TCPListener).Close()
      /tmp/workdir/go/src/net/tcpsock_posix.go:274 +0x49
  crypto/tls.(*listener).Close()
      <autogenerated>:35 +0x7a
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint.func5()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:716 +0xa8
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).onExit.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:429 +0x17a

Previous write by goroutine 83:
  net.ListenTCP()
      /tmp/workdir/go/src/net/tcpsock_posix.go:333 +0x4be
  net.Listen()
      /tmp/workdir/go/src/net/dial.go:395 +0x569
  crypto/tls.Listen()
      /tmp/workdir/go/src/crypto/tls/tls.go:73 +0xc2
  github.com/gravitational/teleport/lib/utils.ListenTLS()
      /gopath/src/github.com/gravitational/teleport/lib/utils/tls.go:46 +0x16f
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:673 +0x9d7
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d

Goroutine 86 (running) created at:
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).onExit()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:431 +0x4c
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:718 +0x1527
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxy.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:577 +0x5b4
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d

Goroutine 83 (running) created at:
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:148 +0x12b
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).Register()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:109 +0x3d6
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).RegisterFunc()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:121 +0x63
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:685 +0x11e0
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxy.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:577 +0x5b4
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d
==================
time="0001-01-01T00:00:00Z" level=info msg="[PROXY] proxy service exited" file="service/service.go:717" 
time="0001-01-01T00:00:00Z" level=info msg="proxySubsys(site=, host=127.0.0.1, port=45992) finished with result: <nil>" file="srv/sshserver.go:737" 
time="0001-01-01T00:00:00Z" level=info msg="[PROXY] web server exited" file="service/service.go:679" 
time="0001-01-01T00:00:00Z" level=info msg="[SUPERVISOR] Service 0x6b4bb0 is done (4)" file="service/supervisor.go:135" 
time="0001-01-01T00:00:00Z" level=info msg="TunClient.Close()" file="auth/tun.go:568" 
time="0001-01-01T00:00:00Z" level=info msg="[AUTH] auth service exited" file="service/service.go:415" 
time="0001-01-01T00:00:00Z" level=info msg="TunClient.Close()" file="auth/tun.go:568" 
time="0001-01-01T00:00:00Z" level=debug msg="TunDialer.Dial(tcp, stub:0)" file="auth/tun.go:605" 
time="0001-01-01T00:00:00Z" level=info msg="SSH server 127.0.0.1:42333 exit
```

klizhentas commented on lib/client/api.go:
```diff
@@ -326,7 +326,9 @@ func (tc *TeleportClient) SSH(command []string, runLocally bool) error {
 
 // Join connects to the existing/active SSH session
-func (tc *TeleportClient) Join(sessionID session.ID) (err error) {
+func (tc *TeleportClient) Join(sid string) (err error) {
```
session ID's are enforced to have `TimeUUID` format, so not sure why you need this?

kontsevoy replied:
it's a re-typed string, causing a ton of typecasts everywhere else. so you're not gaining anything except making caller's life continuously difficult.

klizhentas replied:
it's not a simple string - it's a time UUID formatted ID, so these checks are necessary

klizhentas replied:
you are not supposed to pass any random strings there any ways, you'd need to use session.NewID()

kontsevoy replied:
The session ID is always provided _by a user via UI_ for `Join()` implementation. That's why it makes no sense to require `session.ID` type

klizhentas commented on lib/reversetunnel/agent.go:
```diff
@@ -361,6 +357,7 @@ func (a *Agent) startHeartbeat(conn ssh.Conn) {
 			errC <- err
 			return
 		}
+		hb.SendRequest("ping", false, nil)
```
why do you need this right after the request was sent?

kontsevoy replied:
this speeds up integration tests because it shoots the first ping without waiting

klizhentas replied:
but that's not what going on at least from what I read - in this particular case it simply sends ping twice

klizhentas replied:
see the line above - there's already such request executed, and your line adds one more

klizhentas commented on lib/service/service.go:
```diff
@@ -394,12 +396,12 @@ func (process *TeleportProcess) initAuthService() error {
 		}
 		for !askedToExit {
+			sleepTime := defaults.ServerHeartbeatTTL/2 + utils.RandomDuration(defaults.ServerHeartbeatTTL/10)
+			time.Sleep(sleepTime)
```
what's the purpose of this change? looks like now you start with sleep, while before it was the opposite, just not sure what does it affect really

kontsevoy replied:
see comments. it removes a constant "1 second" delay, replacing it with a flexible value from defaults (thus speeding up tests) and skips the first always-failing attempt at connecting a server which hasn't started yet.

klizhentas replied:
I don't see it (at least looking at the left side of this diff) - to me it views as simple code reshuffling. May be you did not push something?

klizhentas replied:
ah, my bad I see it now, nevermind

klizhentas commented on lib/reversetunnel/agentpool.go:
```diff
@@ -107,10 +107,16 @@ func (m *AgentPool) FetchAndSyncAgents() error {
 		case <-m.closeBroadcast.C:
 			m.Infof("closing")
+			m.Lock()
+			defer m.Unlock()
```
AFAIK this introduces deadlock. as deferred function will be executed once function returns,not at the scope return, this means that any subsequent calls will block forever

kontsevoy committed:
Removed -race flag for integration tests

kontsevoy committed:
PR comments:

Removed -race flag for integration tests
Removed uplicate "ping" in reverse tunnel agent

kontsevoy committed:
Merge branch 'ev/telecast' of github.com:gravitational/teleport into ev/telecast

kontsevoy commented:
ok, updated the branch with PR comments

kontsevoy committed:
Added session id client-side parsing

Booooo!

klizhentas commented:
lgtm
